moveit_planners/ompl/ompl_interface/src/detail/ompl_constraints.cpp

```diff
@@ -0,0 +1,318 @@
+
+  // extract target position and orientation
+  geometry_msgs::Point position =
+      constraints.position_constraints.at(0).constraint_region.primitive_poses.at(0).position;
+  target_position_ << position.x, position.y, position.z;
+  tf::quaternionMsgToEigen(constraints.position_constraints.at(0).constraint_region.primitive_poses.at(0).orientation,
+                           target_orientation_);
+
+  link_name_ = constraints.position_constraints.at(0).link_name;
+  ROS_INFO_STREAM_NAMED(LOGNAME, "Position constraints applied to link: " << link_name_);
+}
+
+void EqualityPositionConstraint::function(const Eigen::Ref<const Eigen::VectorXd>& joint_values,
+                                          Eigen::Ref<Eigen::VectorXd> out) const
+{
+  Eigen::Vector3d error =
+      target_orientation_.matrix().transpose() * (forwardKinematics(joint_values).translation() - target_position_);
+  for (std::size_t dim{ 0 }; dim < 3; ++dim)
+  {
+    if (is_dim_constrained_[dim])
+      out[dim] = error[dim];  // equality constraint dimension
+    else
+      out[dim] = 0.0;  // unbounded dimension
+  }
+}
+
+void EqualityPositionConstraint::jacobian(const Eigen::Ref<const Eigen::VectorXd>& joint_values,
+                                          Eigen::Ref<Eigen::MatrixXd> out) const
+{
+  out.setZero();
+  Eigen::MatrixXd jac = target_orientation_.matrix().transpose() * robotGeometricJacobian(joint_values).topRows(3);
+  for (std::size_t dim{ 0 }; dim < 3; ++dim)
+  {
+    if (is_dim_constrained_[dim])
+      out.row(dim) = jac.row(dim);  // equality constraint dimension
+  }
+}
+
+/************************************
+ * MoveIt constraint message parsing
+ * **********************************/
+std::vector<Bounds> positionConstraintMsgToBoundVector(const moveit_msgs::PositionConstraint& pos_con)
+{
+  auto dims = pos_con.constraint_region.primitives.at(0).dimensions;
+
+  // dimension of -1 signifies unconstrained parameter, so set to infinity
+  for (auto& dim : dims)
+  {
+    if (dim == -1)
+      dim = std::numeric_limits<double>::infinity();
+  }
+
+  return { { -dims[0] / 2, dims[0] / 2 }, { -dims[1] / 2, dims[1] / 2 }, { -dims[2] / 2, dims[2] / 2 } };
+}
+
+/******************************************
+ * OMPL Constraints Factory
+ * ****************************************/
+std::shared_ptr<BaseConstraint> createOMPLConstraint(robot_model::RobotModelConstPtr robot_model,
+                                                     const std::string& group,
+                                                     const moveit_msgs::Constraints& constraints)
+{
+  std::size_t num_dofs = robot_model->getJointModelGroup(group)->getVariableCount();
+  std::size_t num_pos_con = constraints.position_constraints.size();
+  std::size_t num_ori_con = constraints.orientation_constraints.size();
+
+  // Only positions constraints are actually created below.
+  // All the other code is just given feedback to the user.
```

Comment:
?

Reply:
The whole factory method is a template code to support different constraints, but only a single option is actually returns a non-nullptr at the moment (position constraints). All the other cases spit out warnings or errors.

I guess I can remove all the unnecessary template code and only add it when needed, I never though about that.

Reply:
I updated the comment and left the rest of the function as is. I feel it is pretty clear with the log messages and allows for easy integration of orientation constraints and a combination of position and orientation constraints in the future.

Now the comment reads:
```C++
// The factory method contains template code to support different constraints, but only position constraints are
// currently supported. The other options return a nullptr for now and should not be used.
```